test/functional/users_controller_test.rb

```diff
@@ -364,6 +365,42 @@ class UsersControllerTest < ActionController::TestCase
     assert_response :unprocessable_entity
   end
 
+  context "when user is logged in" do
+    setup do
+      @previous_url = "/bookmarks"
+      get :login, set_session_user
+      request.env['HTTP_REFERER'] = @previous_url
+    end
+
+    test "#login redirects to previous url" do
+      get :login
+
+      assert_redirected_to @previous_url
+    end
+
+    test "#login if referer absent redirect to hosts_path" do
+      request.env['HTTP_REFERER'] = nil
+
+      get :login
+
+      assert_redirected_to hosts_path
+    end
+  end
+
+  context "when user has logged out" do
```

Comment:
I think you can remove this block, it tests the same thing as "#login renders login page" (rendering with an empty session).